botorch/acquisition/scorebo.py

```diff
@@ -0,0 +1,186 @@
+    def __init__(
+        self,
+        model: SaasFullyBayesianSingleTaskGP,
+        optimal_outputs: Tensor,
+        optimal_inputs: Optional[Tensor] = None,
+        X_pending: Optional[Tensor] = None,
+        distance_metric: str = "hellinger",
+        estimation_type: str = "LB",
+        sampler: Optional[MCSampler] = None,
+        num_samples: int = 64,
+        maximize: bool = True,
+    ) -> None:
+        r"""Self-correcting Bayesian optimization [hvarfner2023scorebo]_ acquisition
+        function. SCoreBO seeks to find accurate hyperparameters during the course
+        of optimization by incorporating an active learning-like objective into
+        optimization.
+
+        Args:
+            model: A fully bayesian model single-outcome model.
+            optimal_inputs: A `num_samples x num_models x d`-dim tensor containing
+                the sampled optimal inputs of dimension `d`.
+            optimal_outputs: A `num_samples x num_models x 1`-dim Tensor containing
+                the optimal objective values.
+            X_pending: A `batch_shape, m x d`-dim Tensor of `m` design points
+            distance_metric (str, optional): The distance metric used. Defaults to
+                "hellinger".
+            estimation_type: estimation_type: A string to determine which entropy
+                estimate is computed: Lower bound" ("LB") or "Monte Carlo" ("MC")
+                (not implemented yet).
+                Lower Bound is recommended due to the comparable empirical performance.
+            sampler: MCSampler for Monte Carlo estimation of the statistical distance
+                (not implemented yet).
+            num_samples (int, optional): Number of samples if employing monte carlo
+                estimation of the statistical distance. Defaults to 64.
+        """
+
+        super().__init__(model=model)
+        if sampler is None:
+            sampler = SobolQMCNormalSampler(sample_shape=torch.Size([num_samples]))
+        MCSamplerMixin.__init__(self, sampler=sampler)
+
+        # To enable fully bayesian GP conditioning, we need to unsqueeze
+        # to get num_optima x num_gps unique GPs
+        self.maximize = maximize
+        if not self.maximize:
+            optimal_outputs = -optimal_outputs
+
+        # inputs come as num_optima_per_model x num_models x d
+        # but we want it four-dimensional to condition one per model.
+
+        self.optimal_outputs = optimal_outputs.unsqueeze(-2)
+        # JES-like version of SCoreBO if optimal inputs are provided
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore")
+            with fantasize_flag():
+                with settings.propagate_grads(False):
+                    # We must do a forward pass one before conditioning.
+                    self.model.posterior(
+                        self.model.train_inputs[0], observation_noise=False
+                    )
+
+        if optimal_inputs is not None:
```

Comment:
Should the above `model.posterior` call move in here since it is unnecessary otherwise? Only if GPyTorch just did that under the hood rather than requiring a forward pass before...